### PR TITLE
Add expression size limit to parser

### PR DESCRIPTION
**Pull request by jcking**

### CODE REVIEW

File: parser/options.go [hunk 1]
```diff
@@ -49,6 +50,16 @@ func ErrorRecoveryLimit(errorRecoveryLimit int) Option {
 	}
 }
 
+func ExpressionSizeCodePointLimit(expressionSizeCodePointLimit int) Option {
+	return func(opts *options) error {
+		if expressionSizeCodePointLimit < -1 {
+			return fmt.Errorf("error recovery limit must be greater than or equal to -1: %d", expressionSizeCodePointLimit)
```
Comment (TristonianJones) [inline parser/options.go]: Copy paste bug here, I think.

File: parser/parser.go [hunk 2]
```diff
@@ -85,7 +91,14 @@ func (p *Parser) Parse(source common.Source) (*exprpb.ParsedExpr, *common.Errors
 	if !ok {
 		buf = runes.NewBuffer(source.Content())
 	}
-	e := impl.parse(buf, source.Description())
+	var e *exprpb.Expr
+	if buf.Len() > p.expressionSizeCodePointLimit {
+		e = impl.reportError(common.NoLocation,
+			"expression size exceeds limit: size: %d, limit %d",
```
Comment (TristonianJones) [inline parser/parser.go]: Maybe say `codepoint limit`? I think I missed this on the Java review.

File: parser/parser_test.go [hunk 3]
```diff
@@ -1223,3 +1223,17 @@ func TestParse(t *testing.T) {
 		})
 	}
 }
+
+func TestExpressionSizeCodePointLimit(t *testing.T) {
```
Comment (TristonianJones) [inline parser/parser_test.go:1227]: Can you also add a test for a negative valued limit?

**Review by TristonianJones — CHANGES_REQUESTED** [review]

File: parser/parser.go (hunk 2 above)
Comment (jcking) [reply parser/parser.go]: Done.

File: parser/options.go (hunk 1 above)
Comment (jcking) [reply parser/options.go]: Yep, thanks for catching.

File: parser/parser_test.go (hunk 3 above)
Comment (jcking) [reply parser/parser_test.go:1227]: Done.

File: parser/parser_test.go [hunk 4]
```diff
@@ -1223,3 +1223,21 @@ func TestParse(t *testing.T) {
...
+	p, err := NewParser(Macros(AllMacros...), ExpressionSizeCodePointLimit(-2))
+	if err == nil {
+		t.Fatal("expected NewParser to return an error when given an invalid expression size code point limit")
```
Comment (TristonianJones) [inline parser/parser_test.go]: Can you include the `error` in the test failure?

File: parser/parser_test.go [hunk 5]
```diff
@@ -1223,3 +1223,21 @@ func TestParse(t *testing.T) {
...
+	}
+	p, err = NewParser(Macros(AllMacros...), ExpressionSizeCodePointLimit((2)))
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, errs := p.Parse(common.NewTextSource("foo"))
+	if got, want := len(errs.GetErrors()), 1; got != want {
+		t.Fatalf("expected parsing to result in exactly %d errors, got %d errors", want, got)
```
Comment (TristonianJones) [inline parser/parser_test.go]: Usually, you want to phrase the error messages as `got X, wanted Y` rather than the other way around. Here and below.

**Review by TristonianJones — APPROVED** [review]
If you could just address the cosmetic changes in the test error messages, the rest looks ready to submit.

File: parser/parser_test.go (hunk 4 above)
Comment (jcking) [reply parser/parser_test.go]: It's nil, as were are testing that an error did occur. Regardless changed the message.

File: parser/parser_test.go (hunk 5 above)
Comment (jcking) [reply parser/parser_test.go]: Done.
